### PR TITLE
created entities for role, permission, resource and their associations

### PR DESCRIPTION
**Pull request by muthu1**
- Add role, permission, resource entities.
- Change user id to UUID string

### CODE REVIEW

File: src/entities/permission-resource.association.entity.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      permissionId: string,
+      resourceId: string,
+      lastUpdatedBy: string,
+    ) {
+      this.permissionId = permissionId;
+      this.resourceId = resourceId;
+      this.lastUpdatedBy = lastUpdatedBy;
+    }
+  
+    @Column()
```
Comment (m-nikhil) [inline src/entities/permission-resource.association.entity.ts:25]: Index permissionId, else we will scan all the rows

File: src/entities/permission.entity.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    Column,
+    PrimaryGeneratedColumn,
+    CreateDateColumn,
+    UpdateDateColumn,
+    DeleteDateColumn,
+    Index,
+  } from 'typeorm';
+  
+  /**
+   * Represents the Permission that a user can have.
```
Comment (m-nikhil) [inline src/entities/permission.entity.ts:12]: Better,  Represents the Permission that  **role** can have.

Comment (m-nikhil) [reply src/entities/permission.entity.ts:12]: User can have only roles and not permission for now

File: src/entities/permission.entity.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      this.name = name;
+      this.description = description;
+      this.lastUpdatedBy = lastUpdatedBy;
+    }
+  
+    @PrimaryGeneratedColumn('uuid')
+    id: string;
+  
+    @Column()
+    @Index({ unique: true })
```
Comment (m-nikhil) [inline src/entities/permission.entity.ts:30]: is it unique needed? we can maintain at the application level?

As if we delete an entry, it will be marked as deleted  ( deletedAt is punched with time).
Then we can't have a new entry with the same name as a deleted entry. 

Problem of soft delete. Any thoughts?

File: src/entities/resource.entity.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      this.endpoint = endpoint;
+      this.method = method;
+      this.lastUpdatedBy = lastUpdatedBy;
+    }
+  
+    @PrimaryGeneratedColumn('uuid')
+    id: string;
+  
+    @Column()
+    @Index({ unique: true })
```
Comment (m-nikhil) [inline src/entities/resource.entity.ts:34]: name, service, endpoint & method should be a single composite key. 
Not index, separately.

File: src/entities/permission-resource.association.entity.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    Entity,
+    Column,
+    CreateDateColumn,
+    UpdateDateColumn,
+    DeleteDateColumn,
+    Index,
+  } from 'typeorm';
+  
+  /**
+   * Represents the Roles that a user can have.
```
Comment (m-nikhil) [inline src/entities/permission-resource.association.entity.ts:11]: fix the doc, its UserRoleAssociation table

File: src/entities/user-role.association.entity.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    Entity,
+    Column,
+    CreateDateColumn,
+    UpdateDateColumn,
+    DeleteDateColumn,
+    Index,
+  } from 'typeorm';
+  
+  /**
+   * Represents the Roles that a user can have.
```
Comment (m-nikhil) [inline src/entities/user-role.association.entity.ts:11]: fix doc

File: src/entities/user-role.association.entity.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      userId: string,
+      roleId: string,
+      lastUpdatedBy: string,
+    ) {
+      this.userId = userId;
+      this.roleId = roleId;
+      this.lastUpdatedBy = lastUpdatedBy;
+    }
+  
+    @Column()
```
Comment (m-nikhil) [inline src/entities/user-role.association.entity.ts:25]: compositekey userID and roleID

File: dev-database.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
 
 const Docker = require('dockerode');
-const docker = new Docker({socketPath: '/var/run/docker.sock'});
+docker = new Docker({socketPath: '/var/run/docker.sock'});
```
Comment (m-nikhil) [inline dev-database.js:3]: why? revert this change.

File: src/entities/permission-resource.association.entity.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  UpdateDateColumn,
+  DeleteDateColumn,
+  Index,
+} from 'typeorm';
+
+/**
+ * Represents the resources that permissions can have.
+ */
+@Entity()
+@Index(['permissionId', 'resourceId'], { unique: true })
```
Comment (m-nikhil) [inline src/entities/permission-resource.association.entity.ts:15]: I think we can remove all index from all entities as its redunctant, as primary columns will result in composite key which is indexed and unique.

File: src/entities/permission-resource.association.entity.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+export class PermissionResourceAssociation {
```
Comment (m-nikhil) [inline src/entities/permission-resource.association.entity.ts:16]: remove constructors from all entities except the user entity, as it won't be used. I am planning to remove the user also in the next PR.

File: src/entities/role-permission.association.entity.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  UpdateDateColumn,
+  DeleteDateColumn,
+  Index,
+} from 'typeorm';
+
+/**
+ * Represents the Permissions that roles have access to.
+ */
+@Entity()
+@Index(['permissionId', 'roleId'], { unique: true })
```
Comment (m-nikhil) [inline src/entities/role-permission.association.entity.ts:15]: I think we can remove all indexes as its redundant, as primary columns will result in a composite key which is indexed and unique.

File: src/entities/user-role.association.entity.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  UpdateDateColumn,
+  DeleteDateColumn,
+  Index,
+} from 'typeorm';
+
+/**
+ * Represents the Roles that a user can have.
+ */
+@Entity()
+@Index(['userId', 'roleId'], { unique: true })
```
Comment (m-nikhil) [inline src/entities/user-role.association.entity.ts:15]: I think we can remove all indexes as its redundant, as primary columns will result in a composite key which is indexed and unique.

File: dev-database.js (hunk 8 above)
Comment (muthu1) [reply dev-database.js:3]: oh i think some issue while doing git stash...

Comment (muthu1) [reply dev-database.js:3]: will revert it...
